### ISSUE TITLE
Glassfish Upgrade: issues with javascript and  "//<![CDATA[" and "//]]>"

### ISSUE THREAD

**Issue by pdurbin** [IQSS/dataverse#6463]
Split off from #6230 

First the issues:
• user page - if you don't start on API token tab, clicking on that tab is blank
• user page - if you don't start on My Data tab, clicking on that tab is blank
• dataset page - click on versions tab, nothing happens
• dataset page - selecting files on files table, links broken in "## files are currently selected.   Select all ## files in this dataset.  Clear selection." msg above table
• dataset page - selecting edit dataset does nothing
• edit files page - upload doesn't add file to list; if you add a 2nd file, 1st file gets added (true for n+1 and n)
• create Dataset - if validation errors, block ui does not get removed, errors do not appear
• dataset pg + file pg, clicking tabs, in FF (68.2.0) multiple errors in console log: `XML Parsing Error: mismatched tag. Expected: </update>`

Basically, all these issues are caused by XML Parsing error due to the "//<![CDATA[" and closing "//]]>" on various pages. If we remove them, things work again. For example, remove them from my_data_fragment.xhtml and that page works again. remove them from dataset-versions.xhtml and that fixes not only the versions tab, but also the files table links.

That said, I'm not sure removing them is the exact solution we want. They are there to help XML parsing of javascript on xhtml pages (if they contain certain characters, e.g. "<" or "&"). But once we figure out how we want to solve one of these, we can solve all of them (*).

**Comment by scolapasta** [comment]
OK, so I did a lot of research on this last night. tldr is that this seems to be an issue with JSF (solved in 2.3.13), but that there are also others ways to mitigate this.

More detail:
This error ```XML Parsing Error: mismatched tag. Expected: </update>.``` lead me to it being encountered by liferay: https://issues.liferay.com/browse/FACES-3243 

In their "workaround" fix, they "Remove all the "<![CDATA[" and "]]>" tokens since they will interfere with the JSF partial-response."
https://github.com/liferay/liferay-faces-portal/pull/241/files

From there I was eventually able to find the issue tracking it in mojarra:
https://github.com/eclipse-ee4j/mojarra/issues/4392

This issue was fixed in 2.3.13. They describe the issue about user input "]]>", but it still seems to fit our case. We do also have other options:

• we could remove the CDATA tags completely; generally our javascript doesn't have characters that would cause xml issues, but this seems to be against best practice of having these tags
• we can move the javascript to dedicated js files. This seems good practice generally, but I'm not sure we can do that with the jquery ```$(document).ready(function() {...}```

The ideal, of course, would be to use a version of Payara that has uses JSF 2.3.13. @smillidge if you don't mind me pinging you here, is that coming any time soon? (I looked briefly and from what I could tell Payara is still on 2.3.9)

Otherwise we could try to manually upgrade just jSF (seemed non trivial from a quick trial), or do some combination of the above bullets (e.g. move all the functions to js files and remove CDATA around the .ready calls).
 
(we could also try to continue using JSF 2.2, but that would require the same process as going to 2.3.13 and am more inclined to b e forward moving).

**Comment by pdurbin** [comment]
> we could also try to continue using JSF 2.2

Just a heads up that I documented my (failed) attempt at keeping Dataverse on JSF 2.2 (for now) on Payara 5.194 at https://github.com/IQSS/dataverse/issues/6230#issuecomment-568045644

**Comment by poikilotherm** [comment]
If you opt moving to JSF 2.3 (which I second), please keep in mind to upgrade Omnifaces, too. JSF 2.3 compatible version is Omnifaces 3.x. Of course using older versions should work, too, but all the goodies are in that new version, right?

**Comment by smillidge** [comment]
@pdurbin raise an issue on Payara GitHub for the JSF upgrade I suspect nobody has realised that there has been a couple of mojarra releases.

**Comment by pdurbin** [comment]
@smillidge sorry, I'm confused. Are you saying that the upgrade from JSF 2.2 to 2.3 for a Java EE 7 app deployed to Payara 5.194 is unintentional?

**Comment by scolapasta** [comment]
@pdurbin I read this as @smillidge referring to my question about Payara shipping with the more recent version of JSF 2.3 (2.3.13).

@smillidge I am in the process of writing up that Payara github issue as we speak. I did have a talk this week with Veronica Harding from Payara, who was going to check with the tech team for the answer, and creating the github issue for more general awareness would be good, too.

Unless it's something that will ship with a version of Payara being released very soon, we will probably have to go about with a different solution. I plan on going back to the "use a different version than what is shipped", and figuring out why that did not work for either @pdurbin or myself.

**Comment by sekmiller** [comment]
On the dataset page an issue was introduced by moving the javascript around. See #6504 for notes on fixing it. It is unrelated to the CDATA issue inasmuch as removing the tag doesn't break anything.

**Comment by smillidge** [comment]
Sorry yes I was referring to Payara not shipping with the latest patch to Mojarra

**Comment by scolapasta** [comment]
@smillidge great! I did open that issue: https://github.com/payara/Payara/issues/4424. I also tired successfully replaced the jakarta.faces.jar in glassfish/modules and have confirmed that the latest version of JSF does fix the issue.

**Comment by scolapasta** [comment]
Note on your dev environment to use JSF 2.3.14, download from: https://repo1.maven.org/maven2/org/glassfish/jakarta.faces/2.3.14/jakarta.faces-2.3.14.jar

Then replace payara5/glassfish/modules/jakarta.faces.jar (or glassfish5/glassfish/modules/jakarta.faces.jar) with this file (with, of course, the name jakarta.faces.jar)

stop and restart domain and it should be using the new jsf

**Comment by poikilotherm** [comment]
Looks like the fix is gonna make it into 5.201, see payara/Payara#4492

**Comment by scolapasta** [comment]
I'm going to go ahead and close this issue as we are now testing on the latest Payara and this is no longer an issue. (the general payara issue #6230 is still open)
